### PR TITLE
[RFC] Provide override functions

### PR DESCRIPTION
**Pull request by jfthuillier**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | none
| License       | MIT
| Doc PR        | Will do after discussing

This PR is about providing some dev-friendly functions to make overriding cleaner.

Here is an example of how these functions could be used : 

_App.js_ : 
```
import React from 'react';
import { HydraAdmin} from '@api-platform/admin';
import overrideResources from '@api-platform/admin/lib/override/functions';
import parseHydraDocumentation from '@api-platform/api-doc-parser/lib/hydra/parseHydraDocumentation';
import newResources from './override/example';

const entrypoint = process.env.REACT_APP_API_ENTRYPOINT;

const myApiDocumentationParser = entrypoint => parseHydraDocumentation(entrypoint)
  .then(({ api }) => {
    overrideResources(api.resources, newResources);
    return { api };
  })
;

export default () => <HydraAdmin apiDocumentationParser={myApiDocumentationParser} entrypoint={entrypoint} />;
```

So, now the developer has to provide the `newResources` : 
```import React from 'react';
import {TextInput, UrlField} from 'react-admin';
import BookList from '../books/list-1';
import BookCreate from '../books/create-2';
import BookEdit from '../books/edit-1';

const greetingsNameInput = props => (
  <TextInput label={'POC'} {...props} />
);

const greetingsNameField = props => (
  <UrlField {...props} />
);

const booksDescriptionInput = props => (
  <TextInput label={'POC'} {...props} />
);

const greetingsOverride = {
  name: 'greetings',
  fields: [
    {
      name: 'name',
      input: greetingsNameInput,
      field: greetingsNameField,
    }
  ],
  listFields: [],
};

const booksOverride = {
  name: 'books',
  list: BookList,
  create: BookCreate,
  edit: BookEdit,
  fields: [
    {
      name: 'description',
      input: booksDescriptionInput,
    }
  ]
};

export default [
  greetingsOverride,
  booksOverride,
];
```

And here are some examples of what overriding operations could look like : 

A Create : 

```
import React, {Component} from 'react';
import {Create, SimpleForm} from 'react-admin';
import {getResourceField} from '@api-platform/admin/src/override/functions';

/* Rearrange fields : order or div containers */

class BookCreate extends Component {
  render() {
    const {options: {inputFactory, resource}} = this.props;

    return (
      <Create {...this.props}>
        <SimpleForm>
          <div className="row">
            <div className="col-md-6">
              {inputFactory(getResourceField(resource, 'name'))}
            </div>
            <div className="col-md-6">
              {inputFactory(getResourceField(resource, 'description'))}
            </div>
          </div>
        </SimpleForm>
      </Create>
    );
  }
}

export default BookCreate;
```

And an Edit : 

```
import React, {Component} from 'react';
import {Edit, SimpleForm} from 'react-admin';
import {connect} from 'react-redux';
import {formValueSelector} from 'redux-form';
import {getResourceField} from '@api-platform/admin/src/override/functions';

/* Dynamic display */

class BookEdit extends Component {
  render() {
    const {options: {inputFactory, resource}, formValueName} = this.props;

    return (
      <Edit {...this.props}>
        <SimpleForm>

          {inputFactory(getResourceField(resource, 'name'))}

          {formValueName && (
            inputFactory(getResourceField(resource, 'description'))
          )}

        </SimpleForm>
      </Edit>
    );
  }
}

const mapStateToProps = state => ({
  formValueName: formValueSelector('record-form')(state, 'name'),
});

export default connect(mapStateToProps)(BookEdit);
```

Do you think there is any point of include these functions into API-P or this is something that should be done inside the project using API-P ?

### CODE REVIEW

**Comment by dunglas** [comment]
I like the idea!

File: src/override/functions.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
+export const getResource = (resources, resourceName) =>
```
Comment (dunglas) [inline src/override/functions.js]: However I'm not a fond of the `override` name.

Can't we just move this file in file called `docUtils` in the root?

File: src/override/functions.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  resources.find(({name}) => resourceName === name);
+
+export const getResourceField = (resource, resourceFieldName) =>
```
Comment (dunglas) [inline src/override/functions.js]: should be `export function` right? (same for all functions in this file)

File: src/override/functions.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  resource.fields.find(({name}) => resourceFieldName === name);
+
+export const overrideInput = (resource, resourceFieldName, overrideInput) => {
+  const resourceField = getResourceField(resource, resourceFieldName);
+
+  if (resourceField === undefined) {
```
Comment (dunglas) [inline src/override/functions.js]:
```suggestion
  if (undefined === resourceField) return;
```

We use Yoda style, and we inline `if` if there is only one line in the block.

Same everywhere

File: src/override/functions.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
+export const getResource = (resources, resourceName) =>
+  resources.find(({name}) => resourceName === name);
+
+export const getResourceField = (resource, resourceFieldName) =>
+  resource.fields.find(({name}) => resourceFieldName === name);
+
+export const overrideInput = (resource, resourceFieldName, overrideInput) => {
```
Comment (dunglas) [inline src/override/functions.js]: `replaceInput`?

File: src/override/functions.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  const resourceField = getResourceField(resource, resourceFieldName);
+
+  if (resourceField === undefined) {
+    return;
+  }
+
+  resourceField.input = overrideInput;
+};
+
+export const overrideField = (resource, resourceFieldName, overrideField) => {
```
Comment (dunglas) [inline src/override/functions.js]: `replaceField`?

File: src/override/functions.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  const resourceField = getResourceField(resource, resourceFieldName);
+
+  if (resourceField === undefined) {
+    return;
+  }
+
+  resourceField.field = overrideField;
+};
+
+export const overrideResource = (resources, overrideResource) => {
```
Comment (dunglas) [inline src/override/functions.js]: `replaceResource`?

File: src/override/functions.js (hunk 1 above)
Comment (jfthuillier) [reply src/override/functions.js]: Done

File: src/override/functions.js (hunk 2 above)
Comment (jfthuillier) [reply src/override/functions.js]: Done. But just curious, why using `export function` instead of `export const` an arrow function ?

**Comment by dunglas** [comment]
Thanks @jfthuillier! Don't forget to update the docs to explain how to use these new helpers, or they will be very difficult to discover.

**Comment by comxd** [comment]
Thanks @jfthuillier! For the next doc update, here a minimal sample for the "List" view:

```js
import React, {Component} from 'react';
import {
  Datagrid,
  List,
} from 'react-admin';
import {getResourceField} from '@api-platform/admin/src/docsUtils';

/* Rearrange fields : order or div containers */

class PersonPhotoList extends Component {
  render() {
    const {options: {fieldFactory, resource}} = this.props;

    return (
      <List {...this.props}>
        <Datagrid>
          {fieldFactory(getResourceField(resource, 'name'))}
        </Datagrid>
      </List>
    );
  }
}

export default PersonPhotoList;
```

**Pull request merged**